zephyr/Kconfig

```diff
@@ -0,0 +1,680 @@
+# Copyright (c) 2017-2020 Linaro Limited
+# Copyright (c) 2020 Arm Limited
+# Copyright (c) 2022 Legrand North America, LLC.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+
+# Option to build the project with the MCUBoot application
```

Comment:
This now appears at `Modules → mcuboot` in menuconfig instead of at the top level... mcuboot isn't a module, it's an application. I don't find this to be an enhancement, and this is going to confuse people that have been used to mcuboot, the application, being at the top level, where other applications and zephyr best practice states to put it at.

After having a play with it, I don't understand this change. If I configure hello_world without sysbuild, I see these options - why? I can't enable mcuboot from here, and if I do enable these options, all I get is a broken build with compile errors about `The flash driver lacks page layout support; falling back on hacks.` which doesn't make sense. And if I configure the application using sysbuild with mcuboot, these settings aren't even set, again, if I enable them, the build suddenly breaks with the same error.

Reply:
From [Modules (External projects)](https://docs.zephyrproject.org/latest/develop/modules.html):

> Zephyr relies on the source code of several externally maintained projects in order to avoid reinventing the wheel and to reuse as much well-established, mature code as possible when it makes sense. In the context of Zephyr’s build system those are called modules. 

And the only specific requirement on all modules is:

> The module repository codebase shall include a module.yml file in a zephyr/ folder at the root of the repository.

@nordicjm wrote:
> mcuboot isn't a module, ...

Based on this definition and criteria, `mcuboot` was a downstream module before I ever started. It just wasn't represented that way in the build system.

> ... it's an application.

Consider:
From [Zephyr's Glossary entry for "application"](https://docs.zephyrproject.org/latest/glossary.html#term-application):

> The set of user-supplied files that the Zephyr build system uses to build an application image for a specified board configuration. It can contain application-specific code, kernel configuration settings, and at least one CMakeLists.txt file.

In order for mcuboot to live up to the reuse intent, the sources are not user-supplied any more than the drivers are user-supplied. 

The use case that prompted this change is my need to control and build boot code for my project without having to fork or change repositories in `zephyrproject-rtos`.  My "application" is simply one CMakeLists.txt, configuration settings files, and one 'keys.c' source file.  (FYI, all my other projects either provide an empty `main(void){}` or link in the weakly defined one provided by the kernel.)

From this perspective, the mcuboot app is a service that can be built into an application, just like a UI, or BLE beacon (or TCP/IP "ping"/"echo"/"data source").

> this is going to confuse people that have been used to mcuboot, the application, being at the top level, where other applications and zephyr best practice states to put it at.

People will be confused adjusting to the Zephyr definition of "application", which does not require any logic.  In fact, Zephyr kernel provides a weakly defined, default `void main(void)` definition, and the Zephyr build system works with the application providing only a zero-byte .c file (no symbols).

The top-level CMakeLists.txt still provides a title for the "application" configuration.

> If I configure hello_world without sysbuild, I see these options - why?

The better question is "For a certifiable image should `main()` contain any application-level functionality?"  For certifiable images every line of code needs to be traced back to requirements and verified that it meets the requirements.  The most efficient way to verify this is with automated unit testing. The Zephyr testing framework provides the definition of `void main(void)`, so the code under test must not. So, though `main()` _can_ contain logic, it probably shouldn't.

Separately, notice that nothing prevents building an image that combines outputing "hello world!" with booting an image - other than that both currently define the symbol "main()".  As we have noticed, neither should define that symbol.

Q: Would you prefer me to change the mcuboot source code to no longer define main() in this PR or in one that introduces a testsuite for main?

>  I can't enable mcuboot from [hello_world sample", and if I do enable these options, all I get is a broken build with compile errors about The flash driver lacks page layout support; falling back on hacks. which doesn't make sense

Actually, the broken build makes perfect sense.  You are trying to build a feature set for a target+drivers that do not support it. Why do you expect it would always work?

> And if I configure the application using sysbuild with mcuboot, these settings aren't even set, again, if I enable them, the build suddenly breaks with the same error.

Sysbuild is a convenience tool that supports building and flashing two images at once.  One _very_ reasonable scenario that should be supported by Sysbuild is to build and reflash an immutable boot image with a mutable one with a different feature set (e.g., one that drives a graphics display).  If Sysbuild can't support this scenario then it is lacking functionality.

Reply:
> From this perspective, the mcuboot app is a service that can be built into an application, just like a UI, or BLE beacon (or TCP/IP "ping"/"echo"/"data source").

No it is not, mcuboot is a bootloader, it provides services/libraries that an application can use and zephyr does, but mcuboot itself is not just something you add onto an application, it is a separate application altogether. Modules in zephyr in Kconfig control features for a build, for the mcuboot repository this would include the libraries that mcuboot includes as these can and are used from an application.

> The better question is "For a certifiable image should main() contain any application-level functionality?"

I really do not care about that, that sort of thinking is for the safety auditable branches of zephyr and other repositories, the main branches used by zephyr should not bend to the need of a safety critical/certifiable need if it is detrimental to the users that do not need that.

> Q: Would you prefer me to change the mcuboot source code to no longer define main() in this PR or in one that introduces a testsuite for main?

I think you are grossly misunderstanding how mcuboot works, mcuboot does and absolutely SHOULD have a main, that's how the bootloader works, it must operate without multi-tasking/threading support. This comment makes no sense about mcuboot.

> Actually, the broken build makes perfect sense. You are trying to build a feature set for a target+drivers that do not support it. Why do you expect it would always work?

So mcuboot doesn't work on nordic devices? Why does it build and work fine when I build mcuboot for these myself, or build an application and mcuboot for the same boards using sysbuild?

> Sysbuild is a convenience tool that supports building and flashing two images at once. One very reasonable scenario that should be supported by Sysbuild is to build and reflash an immutable boot image with a mutable one with a different feature set (e.g., one that drives a graphics display). If Sysbuild can't support this scenario then it is lacking functionality.

Sysbuild can and does support that (as in, sysbuild right now as it is in-tree). We currently have sysbuild configuring multiple images per core for a dual-core MCU (nRF53) and that is working fine.

Reply:
And just to show how simple sysbuild is, I can do this right now in zephyr:
```
cd zephyr/samples/hello_world
mkdir _build
cd _build
cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -DAPP_DIR=.. ../../../share/sysbuild
ninja flash
```

Which will build hello_world and mcuboot for my board, and will then flash both images to my board. I can even change mcuboot's configuration by altering the cmake like with e.g. `-Dmcuboot_CONFIG_LOG=n`

Reply:
> mcuboot is a bootloader, it provides services/libraries that an application can use and zephyr does, but mcuboot itself is not just something you add onto an application, ...

Please restate this following the definition of "application" from the Zephyr glossary. Either mcuboot is or is not an application.  If it can provide services/libraries to other applications, then it is more than an application (just like example-application is a manifest-module rather than just a module.

> the main branches used by zephyr should not bend to the need of a safety critical/certifiable need if it is detrimental to the users that do not need that.

Please review the [Zephyr Project statement,](https://www.zephyrproject.org/wp-content/uploads/sites/38/2020/09/CLEAN-LF-Zephyr-Charter-20200624-effective-20200901.pdf).  Item 1.b states:
> maintain an auditable code base, while taking advantage of community
participation; this auditable code base is open source;

Item 4e [f] states (reformatted, emphasis mine):
> The Security [Safety] Architect shall
> - serve on the TSC as a “security maintainer,” ["safety maintainer"] and 
> - is responsible for verifying that the **Project** is following the processes for selected
standards that the Security [Safety] Committee recommends and
> - that the Project is appropriately documenting features to move from the long term support (“LTS”)
code base to the auditable code base. 
> It is intended that Maintainers will seek the sign-off from the Security [Safety] Architect prior to implementing significant changes to the LTS code base **where such changes may result in significant changes in
functionality between the LTS and auditable code bases**.

The above items state that developers **should** bend to the need of a safety critical/certifiable need unless working on functionality or design that has been granted an exception by both the Safety Architect and the Security Architect.       

> I think you are grossly misunderstanding how mcuboot works

Actually I understand it better than most, having heavily designed and refactored single- and multi-stage boot code for an MCU-based product line for high volume, long-lived devices.

> it must operate without multi-tasking/threading support

That is one of the use cases.  Another requires that that it work when in a multi-tasking/threading environment (e.g., when MCUboot zephyr port is using the USB interface for serial recovery, or any situation a where an end user integrates a multi-threaded solution for any of the support APIs like UI for CGD).

> mcuboot does and absolutely SHOULD have a main,

You are confusing design with implementation.  MCUboot does provide top-level logic that must automatically run once the lower layers are initialized.  The implementation does not have to provide the symbol `main()`.  It could be achieved by providing the symbol `mcuboot_main()` which is called by either Zephyr's `main()`, ztest suite, or the last executed `SYS_INIT()`.

> So mcuboot doesn't work on nordic devices?

Not what I said.  The build error shows that the subset of configuration settings used were insufficient and/or incorrectly set for the target. I never claimed that integrating MCUBOOT_APP into the hello_world sample only required setting the one Kconfig.  

BTW, if you look at the commit comments you will see that I explicitly verified my changes built correctly for nrf52840dk_nrf52840 as well as for all the platforms pulled in by twister for the referenced test suites.

> Sysbuild can and does support that (as in, sysbuild right now as it is in-tree).

As an end user, I need to build and release my boot image independent of building other images. MCUboot needs to support building this straight from CMake/ninja without using west or sysbuild.

Separately, working on MCUboot issue #1499 I found the Zephyr testing framework/structure does not work with the current mcuboot zephyr port build files, but it does work with these changes. This is completely independent of sysbuild.

Reply:
> The above items state that developers should bend to the need of a safety critical/certifiable need unless working on functionality or design that has been granted an exception by both the Safety Architect and the Security Architect.

No it doesn't, it talks about introducing changes into LTS branches, this is nothing to do with LTS.

Reply:
In order to release a certified bootloader the processes and tools to build that must exist in auditable branch, and exceptions to this in LTS should be signed off by the Safety & Security Architects.  Until that is signed off then the feature should exist in LTS too. This includes building artifacts verifying functionality in the mcuboot top-level logic.

Reply:
> this is nothing to do with LTS

I agree that this has nothing to do with introducing changes to the LTS branch, but it does have to do with whether the existing mcuboot build files work with the Zephyr LTS branch build features that meet my user needs.  Sysbuild does not meet my user needs and does not address the need to build correctly with the Zephyr testing structure.